keel-bakery/src/main/kotlin/com/netflix/spinnaker/keel/bakery/resource/ImageHandler.kt

```diff
@@ -0,0 +1,131 @@
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.ResourceKind
+import com.netflix.spinnaker.keel.api.ResourceName
+import com.netflix.spinnaker.keel.api.SPINNAKER_API_V1
+import com.netflix.spinnaker.keel.bakery.api.ImageSpec
+import com.netflix.spinnaker.keel.clouddriver.CloudDriverService
+import com.netflix.spinnaker.keel.events.TaskRef
+import com.netflix.spinnaker.keel.model.Job
+import com.netflix.spinnaker.keel.model.OrchestrationRequest
+import com.netflix.spinnaker.keel.model.OrchestrationTrigger
+import com.netflix.spinnaker.keel.orca.OrcaService
+import com.netflix.spinnaker.keel.persistence.ArtifactRepository
+import com.netflix.spinnaker.keel.plugin.ResolvableResourceHandler
+import com.netflix.spinnaker.keel.plugin.ResourceDiff
+import com.netflix.spinnaker.keel.plugin.ResourceNormalizer
+import kotlinx.coroutines.runBlocking
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+class ImageHandler(
+  override val objectMapper: ObjectMapper,
+  override val normalizers: List<ResourceNormalizer<*>>,
+  private val cloudDriver: CloudDriverService,
+  private val artifactRepository: ArtifactRepository,
+  private val baseImageCache: BaseImageCache,
+  private val orcaService: OrcaService,
+  private val igorService: ArtifactService
+) : ResolvableResourceHandler<ImageSpec, Image> {
+
+  override val apiVersion = SPINNAKER_API_V1.subApi("bakery")
+  override val supportedKind = ResourceKind(
+    apiVersion.group,
+    "image",
+    "images"
+  ) to ImageSpec::class.java
+
+  override fun generateName(spec: ImageSpec): ResourceName =
+    ResourceName("bakery:image:${spec.artifactName}")
+
+  override fun desired(resource: Resource<ImageSpec>): Image =
+    with(resource.spec) {
+      val latestVersion = artifactRepository
+        .versions(DeliveryArtifact(artifactName, DEB))
+        .first()
+      Image(
+        baseAmiVersion = baseImageCache.getBaseImage(baseOs, baseLabel),
+        // TODO: there must be a canonical way of composing this
+        appVersion = latestVersion,
+        regions = regions
+      )
+    }
+
+  override fun current(resource: Resource<ImageSpec>): Image? =
+    runBlocking {
+      cloudDriver.images("aws", resource.spec.artifactName)
+        .await()
+        .firstOrNull()
+        ?.let {
+          // TODO: this is pretty crude to say the least
+          val tags = it.tagsByImageId.values.first()!!
+          Image(
+            tags.getValue("base_ami_version")!!,
+            tags.getValue("appversion")!!,
+            it.amis.keys
+          )
+        }
+    }
+
+  override fun upsert(
+    resource: Resource<ImageSpec>,
+    resourceDiff: ResourceDiff<Image>?
+  ): List<TaskRef> {
+    val taskRef = runBlocking {
+      val (_, application, version) = Regex("([\\w_]+)-(.+?)/.*")
+        .find(resourceDiff!!.source.appVersion)!!.groupValues
+      val artifact = igorService.getArtifact(application, version)
```

Comment:
Before baking fetch the full artifact from Igor.

We _could_ store this when we initially get it via Rocket events but there’s some consensus that we need a mechanism to catch up on any potentially missed events and I kind of don’t want to replicate a ton of data from elsewhere.